Protocol.md

```diff
@@ -0,0 +1,177 @@
+
+#### CET signatures
+A set of signatures from B, one for each CET input.
+
+#### Refund signature
+A signature from B for the refund transaction input.
+
+### Sign
+
+After receiving and [validating](#sign-validation) the content of the `offer` message, A replies with a `sign` message containing:
+1. Fund transaction signatures
+1. CET signatures
+1. Refund signature
+
+#### Fund transaction signatures
+A set of signatures from A, one (in case of P2WPKH) or several (in case of multisig P2WSH) for each of A's input to the fund transaction.
+
+#### CET signatures
+A set of signatures from A, one for each CET input.
+
+#### Refund signature
+A signature from A for the refund transaction input.
+
+### Broadcast
+Upon receiving the `sign` message from A, B [validates](#sign-validation) its content, signs the fund transaction and broadcasts it to the Bitcoin network.
+
+## Validation
+
+This section describes how each party needs to validate the messages received.
+We distinguish between user validation, performed by the user (through a user interface), and machine validation, that can be done without user interaction.
+
+### Offer Validation
+
+#### User validation
+* Contract Information: confirm that the contract information matches what was agreed upon.
+* Oracle information: confirm that the oracle specified in the message matches what was agreed upon.
+* A fund amount: confirm that the input amount matches what was agreed upon.
+* Estimatesmartfee: confirm that it corresponds to what was agreed upon.
+* CET CSV delay: confirm that it corresponds to what was agreed upon.
+* Refund locktime: confirm that it corresponds to what was agreed upon.
+
+#### Machine validation
+
+* A fund amount: given this and B fund amount, need to validate that the amount incorporates the required fees for the CET or refund transaction as well as the closing transaction, and that the total input amount is greater than the payout amounts plus the fees to be paid.
+* A public key: validate that the public key is well formed.
+* Alice inputs: validate that the referenced outputs are in fact unspent and are either P2WSH or P2WPKH.
+
+### Accept validation
+
+#### User validation
+* B fund amount: confirm that the fund amount matches what was agreed upon.
+
+#### Machine validation
+* Bob public key: validate that the public key is well formed.
+* Bob inputs: validate that the referenced outputs are in fact unspent and are either P2WSH or P2WPKH.
+* CET signatures: validate that the signatures are valid for each of the CETs.
+* Refund signature: validate that the signature is valid.
+
+### Sign validation
+
+#### Machine validation
+* Fund transaction signatures: validate that the signatures are valid for each of the provided inputs.
+* CET signatures: validate that the signatures are valid for each of the CETs input.
+* Refund signature: validate that the signature is valid for the refund transaction input.
+
+## Contract cancellation
```

Comment:
We should probably specify a `Cancel` message here, especially in the case of wanting to cancel after sign. This way things are clear, especially since the person cancelling should always be able to outbid the fee of the funding tx which cannot be raised.

Reply:
I'm not sure I see the point of a cancel message after signing (or sending the signature). As written, if a party wants to cancel after that they should try to double spend the funding transaction, but it doesn't really involve the other party AFAICT. And for cancellation prior to that, I think as written an empty message is enough (of course we should later on define some message format and IDs to recognize it, but I feel that is a next step).

Reply:
Essentially I'm saying that whatever message we specify for use in normal cancellation should also be recommended for use when double-spending to cancel. Essentially it is the difference between polite and impolite disconnection in most networks applications, it's always nice to have a way of saying goodbye so the other party can know definitively what happened rather than just hanging up without saying anything. (Certainly there's nothing forcing users to follow this and send the message when doing a double-spend cancel but if out-of-the-box implementations do it then it will provide a better UX overall I think)

Reply:
I've been convinced, there's no real reason to have a cancel message as there is nothing that could have been done in response to there not being one (impolite close)